flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBNativeMetricMonitor.java

```diff
@@ -0,0 +1,154 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.View;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.Closeable;
+import java.math.BigInteger;
+
+/**
+ * A monitor which pulls {{@link RocksDB}} native metrics
+ * and forwards them to Flink's metric group. All metrics are
+ * unsigned longs and are reported at the column family level.
+ */
+@Internal
+public class RocksDBNativeMetricMonitor implements Closeable {
+	private static final Logger LOG = LoggerFactory.getLogger(RocksDBNativeMetricMonitor.class);
+
+	private final RocksDBNativeMetricOptions options;
+
+	private final MetricGroup metricGroup;
+
+	private final Object lock;
+
+	@GuardedBy("lock")
+	private RocksDB rocksDB;
+
+	RocksDBNativeMetricMonitor(
+		@Nonnull RocksDB rocksDB,
+		@Nonnull RocksDBNativeMetricOptions options,
+		@Nonnull MetricGroup metricGroup
+	) {
+		this.options = options;
+		this.metricGroup = metricGroup;
+		this.rocksDB = rocksDB;
+
+		this.lock = new Object();
+	}
+
+	/**
+	 * Register gauges to pull native metrics for the column family.
+	 * @param columnFamilyName group name for the new gauges
+	 * @param handle native handle to the column family
+	 */
+	void registerColumnFamily(String columnFamilyName, ColumnFamilyHandle handle) {
+		MetricGroup group = metricGroup.addGroup(columnFamilyName);
+
+		for (String property : options.getProperties()) {
+			RocksDBNativeMetricView gauge = new RocksDBNativeMetricView(handle, property);
+			group.gauge(property, gauge);
+		}
+	}
+
+	/**
+	 * Updates the value of metricView if the reference is still valid.
+	 */
+	private void setProperty(ColumnFamilyHandle handle, String property, RocksDBNativeMetricView metricView) {
+		try {
+			synchronized (lock) {
+				if (rocksDB != null) {
+					long value = rocksDB.getLongProperty(handle, property);
+					metricView.setValue(value);
+				}
+			}
+		} catch (RocksDBException e) {
+			LOG.warn("Failed to read native metric %s from RocksDB", property, e);
```

Comment:
We've had issues in the past where the amount of logging messages where drowning out other messages.
If this exception occurs, and if multiple metrics (or god forbid all of them) are enabled, then this could lead to a large amount of logging messages.

I suggest to add a flag to each gauge so that this message is only logged once for each of them.